### DISCUSSION TITLE
How to add outline at the material boundary

### DISCUSSION THREAD

**Discussion by HSDSZ** [bnpr/Malt#105] [Q&A] [answered]
![inside outline](https://user-images.githubusercontent.com/36077492/127741800-d2cebd82-a0e6-4519-aae8-c5879656b471.png)

I have a plane. For the left half, I assign a blue color with an outline. For the right half, I assign a red color with an outline. 

The plane has a perfect outline. 
However, the boundary does not have line. Can malt shows the material boudary line?

**Comment by pragma37** [comment, marked as answer]
At the moment, IDs are per object so all materials in the same object share the same ID.
You can, however, set a per-material custom ID.

![imagen](https://user-images.githubusercontent.com/36610999/127746875-32fc136f-2be8-464c-bdcd-668673e04798.png)

```GLSL
#include "Pipelines/NPR_Pipeline.glsl"

uniform vec3 ambient_color = vec3(0.1,0.1,0.1);
uniform vec3 diffuse_color = vec3(1.0,0.1,0.1);
uniform vec3 specular_color = vec3(1.0,1.0,1.0);
uniform float roughness = 0.5;

uniform vec4 line_color = vec4(0.0,0.0,0.0,1.0);
uniform float line_width = 1.0;
uniform float line_depth_threshold = 0.5;
uniform float line_normal_threshold = 1.0;

uniform int custom_id_offset = 0;

void COMMON_PIXEL_SHADER(Surface S, inout PixelOutput PO)
{
    PO.id += custom_id_offset;

    vec3 diffuse = diffuse_color * get_diffuse_half();
    vec3 specular = specular_color * get_specular(roughness);
    vec3 color = ambient_color + diffuse + specular;

    PO.color = vec4(color, 1.0);

    PO.line_color = line_color;
    PO.line_width = get_line_simple(line_width, line_depth_threshold, line_normal_threshold);
}
```

You should use a large value for the custom_id_offset, so it doesn't overlap with other objects.

This is an area that needs a redesign.
Ideally, it should be possible to have per object, mesh, material and custom IDs.

**Reply by HSDSZ** [reply]
Thanks. This works
